### PR TITLE
Refactor chunk index handling

### PR DESCRIPTION
**Pull request by erimatnor**
This change refactors the chunk index handling to make better use
of standard PostgreSQL catalog information, while removing the
hypertable_index metadata table and associated triggers, including
those on the chunk_index table. The chunk_index table itself is
also simplified.

A benefit of this refactoring is that indexes are no longer
created using string mangling to construct the CREATE INDEX command
for a chunk, based on the string definition of the hypertable
index. Instead, indexes are created in C using proper index-related
internal data structures.

Chunk indexes can now also be renamed and are added in the chunk
table's tablespace by default, unless another tablespace is explicitly
specified when creating the index. Changing tablespace on a hypertable
index also recurses to chunks, as expected.

Creating Hypertable indexes with the CONCURRENTLY modifier is
currently blocked, due to unclear semantics regarding concurrent
creation over many tables, including how to deal with snapshots.

### CODE REVIEW

**Comment by erimatnor** [comment]
A couple of notes to help reviewing this:

* Most changes in plpgsql are removals to get rid of the hypertable_index table and related triggers and functions.
* Additions are mostly in `indexing.c`, `chunk_index.c` (new files), and `process_utility.c`
* Large number of changes in tests are mostly due to new index names on chunks (that now follow PostgreSQL's naming convention).
* There's a new index test, specifically for index functionality
* My editor automatically removes whitespaces in modified files, so, unfortunately, some plpgsql code has such non-code logic changes

File: src/process_utility.c [hunk 1]
```diff
@@ -407,19 +414,59 @@ process_drop_trigger(DropStmt *stmt)
...
+static void
 process_drop(Node *parsetree)
 {
 	DropStmt   *stmt = (DropStmt *) parsetree;
 
 	switch (stmt->removeType)
 	{
 		case OBJECT_TABLE:
-			return process_drop_table(stmt);
+			process_drop_table(stmt);
```
Comment (cevian) [inline src/process_utility.c]: Why do you think this style is better? A return clearly indicates that nothing happens after....

File: test/expected/alter.out [hunk 2]
```diff
@@ -26,7 +26,7 @@ SELECT * FROM alter_before;
 -- now different for the root table and the chunk
 SELECT c.relname, a.attname, a.attnum FROM pg_attribute a, pg_class c
 WHERE a.attrelid = c.oid
-AND (c.relname LIKE '_hyper_1%' OR c.relname = 'alter_before')
+AND (c.relname LIKE '_hyper_1%_chunk' OR c.relname = 'alter_before')
```
Comment (cevian) [inline test/expected/alter.out]: Why this change?

File: test/expected/ddl.out [hunk 3]
```diff
@@ -397,14 +387,27 @@ FOR EACH STATEMENT EXECUTE PROCEDURE empty_trigger_func();
 --expect error cases
 \set ON_ERROR_STOP 0
```
Comment (cevian) [inline test/expected/ddl.out]: get rid of  \set ON_ERROR_STOP 0 no longer ERROR

File: test/expected/ddl_single.out [hunk 4]
```diff
@@ -339,14 +329,27 @@ FOR EACH STATEMENT EXECUTE PROCEDURE empty_trigger_func();
 --expect error cases
 \set ON_ERROR_STOP 0
```
Comment (cevian) [inline test/expected/ddl_single.out]: Get rid of ON_ERROR_STOP 0

File: test/expected/index.out [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+INSERT INTO index_test VALUES ('2017-01-20T09:00:01', 17.5);
+-- Index should be in table's tablespace
+\d+ index_test
+                              Table "public.index_test"
+ Column |           Type           | Modifiers | Storage | Stats target | Description 
+--------+--------------------------+-----------+---------+--------------+-------------
+ time   | timestamp with time zone |           | plain   |              | 
+ temp   | double precision         |           | plain   |              | 
+Indexes:
+    "index_test_time_idx" btree ("time")
```
Comment (cevian) [inline test/expected/index.out]: Hmm the index here does not appear in tablespace1. Shouldn't we be consistent with what postgres does?

File: sql/tables.sql [hunk 6]
```diff
@@ -59,7 +59,7 @@ CREATE TABLE IF NOT EXISTS _timescaledb_catalog.tablespace (
 SELECT pg_catalog.pg_extension_config_dump('_timescaledb_catalog.tablespace', '');
 
 -- A dimension represents an axis along which data is partitioned.
-CREATE TABLE _timescaledb_catalog.dimension (
+CREATE TABLE IF NOT EXISTS _timescaledb_catalog.dimension (
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply sql/tables.sql]: First off, this makes things consistent since some tables/objects already had `IF NOT EXISTS`. Second, why shouldn't it be part of upgrade scripts? Can you clarify the reason why not and what makes this file special? Third, this is in fact needed so that one can change the catalog schema in more advanced ways than simply removing/adding columns without having to mirror a full `CREATE TABLE` in the upgrade script, which might be more error prone.

File: Makefile [hunk 7]
```diff
@@ -19,7 +19,7 @@ UPDATE_VERSIONS_PRE = $(shell ls -1 sql/updates/pre-$(UPDATE_VERSIONS).sql)
 UPDATE_VERSIONS_POST = $(shell ls -1 sql/updates/post-$(UPDATE_VERSIONS).sql)
 
 # Get SQL files that are needed to build update script.
-UPDATE_SQL_FILES = $(shell echo ${UPDATE_VERSIONS_PRE} && cat sql/functions_load_order.txt && echo ${UPDATE_VERSIONS_POST} )
+UPDATE_SQL_FILES = $(shell echo ${UPDATE_VERSIONS_PRE} && cat sql/setup_load_order.txt && cat sql/functions_load_order.txt && echo ${UPDATE_VERSIONS_POST} )
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply Makefile]: Is this separation really needed? Why would some things need explicit update commands (I guess you mean ALTER?) and some not? Maybe I am missing something with how this should work? Can you elaborate?

File: src/chunk_index.c [hunk 8]
```diff
@@ -0,0 +1,533 @@
...
+	return indexlist;
+}
+
+/*
+ * Pick a name for a chunk index.
+ *
+ * The goal is that a chunk index name should be analogous to the corresponding
+ * hypertable index name. For instance, if the hypertable's index name is
+ * "<table_name>_<columns>_<suffix>" we should generate the name
+ * "<chunk_table_name>_<columns>_<suffix>".
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk_index.c]: If you are referring to your comment about "Unique1" vs <chunk_name>_<column_names> in a test output below, this function is not responsible for that output.

The reason you get that mismatch is because when you are creating a new index on an existing table, this function is actually not involved, and instead the IndexStmt is simply recursed to chunks, letting PostgreSQL take care of naming indexes and resolving conflicts. 

So, as I see it, there are two approaches to naming indexes:

1) we let PostgreSQL pick names to the extent possible, reusing its code.
2) we always generate our own names using our own code.

I opted for (1). Still, the problem is that the index naming functions are non-public, so the only way to get PostgreSQL to generate a name is to set index name to NULL in the IndexStmt for each chunk. Although this admittedly gives some not ideal corner-case mappings, but it does use PostgreSQL's own naming functions and conflict resolution to the extent possible. 

Alternative (2) probably means using this function or a variant, and maybe that is the approach we should take rather than letting PostgreSQL pick names? I am certainly open for it. Probably <chunk_table>_<original_index_name> is the simplest approach, with some conflict resolution in place. 

For the record though, the previous implementation didn't do any conflict resolution and simply threw an error trace if an index name conflicted on a chunk. Not a likely error, but still possible technically.  So, no matter the naming scheme we end up with this is an improvement from a correctness standpoint.

File: src/process_utility.c (hunk 1 above)
Comment (erimatnor) [reply src/process_utility.c]: This function returns void now.

File: test/expected/index.out (hunk 5 above)
Comment (erimatnor) [reply test/expected/index.out]: This index isn't really used so not sure? But I guess we could use the main index's tablespace and recurse to chunks instead of the hypertable's? The only downside is that you have to create all indexes in the "correct" tablespace. It gives you more control but less convenience. Although I see the correctness standpoint.

File: sql/tables.sql (hunk 6 above)
[the remark this answers is not in the document]
Comment (erimatnor) [reply sql/tables.sql]: Just to be clear, in the upgrade script I rename the old chunk_index table, then let the tables.sql install the new/updated table, then in the post script I convert data from the old table to the new table, finally dropping the old table. This avoids having to specify an identical full CREATE TABLE statement in the upgrade script. It felt messy (if even possible) to ALTER TABLE my way through an update to the table that removes columns and converts  data simultaneously. Maybe it is possible, but it also leaves the table with a lot of "garbage" column definitions. I also think that risks messing up the row index mapping in catalog.c, because of the garbage columns. The only safe ALTER TABLE is to append columns at the end.

Including this file seems similar to what happens with functions on other objects during upgrades (i.e., we include the full source code, irrespective of whether there have been modifications or not).

[the remark this answers is not in the document]
Comment (olofr) [reply sql/tables.sql]: I recall us having lengthy discussions about this and IIRC the conclusion was that we should avoid `IF NOT EXISTS`. I believe one reason was that it will not do any checks on the actual format of the table, only that there exists a table of that name.

[the remark this answers is not in the document]
Comment (erimatnor) [reply sql/tables.sql]: I don't think I was involved in this discussion so I don't know the context. Still, I don't see why the lack of validation when using `IF NOT EXISTS` would be a problem as it would be no worse than before and this is not a mechanism to replace tables anyway (i.e., to replace a table you need first to remove it in an update script). Without including this file in the update, we'd have to specify two identical  `CREATE TABLE` statements in two places. I can certainly do this but I still haven't heard a really good reason for why we are doing things this way. Right now the update script compilation is quite confusing, SQL files are listed across three different `.txt` files and there are no docs describing their function apart from vague naming.

File: test/expected/ddl_single.out (hunk 4 above)
Comment (erimatnor) [reply test/expected/ddl_single.out]: Shouldn't rename column generate error? Can remove, but doesn't seem related to my changes.

**Comment by erimatnor** [comment]
@cevian Addressed most comments, apart from update script that we might have to resolve separately. Notable changes:

- Chunk indexes are now always named <chunk_table_name>_<main_index_name>
- Renaming an index also renames chunk indexes
- Chunk indexes are created in the main index's tablespace as indicated in the CREATE INDEX command
- Default indexes are created in the main table's tablespace

File: src/chunk_index.c [hunk 9]
```diff
@@ -0,0 +1,533 @@
...
+				BTEqualStrategyNumber, F_INT4EQ, Int32GetDatum(chunk->fd.id));
+	ScanKeyInit(&scankey[1], Anum_chunk_index_chunk_id_index_name_idx_index_name,
+				BTEqualStrategyNumber, F_NAMEEQ, CStringGetDatum(indexname));
+
+	return chunk_index_scan_update(CHUNK_INDEX_CHUNK_ID_INDEX_NAME_IDX,
+						  scankey, 2, chunk_index_tuple_rename, &renameinfo);
+}
+
+int
+chunk_index_rename_parent(Hypertable *ht, Oid hypertable_indexrelid, const char *newname)
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply src/chunk_index.c]: I have now implemented this change.
